datafusion/expr/src/partition_evaluator.rs

```diff
@@ -210,18 +229,20 @@ pub trait PartitionEvaluator: Debug + Send {
         ))
     }
 
-    /// Called for window functions that use values from window frame,
-    /// such as `FIRST_VALUE`, `LAST_VALUE`, `NTH_VALUE` and produce a
-    /// single value for every row in the partition.
+    /// Does the window function use the values from its window frame?
     ///
-    /// Returns a [`ScalarValue`] that is the value of the window function for the entire partition
-    fn evaluate_inside_range(
-        &self,
-        _values: &[ArrayRef],
-        _range: &Range<usize>,
-    ) -> Result<ScalarValue> {
-        Err(DataFusionError::NotImplemented(
-            "evaluate_inside_range is not implemented by default".into(),
-        ))
+    /// If this function returns true, [`Self::create_evaluator`] must
+    /// implement [`PartitionEvaluator::evaluate_inside_range`]
+    fn uses_window_frame(&self) -> bool {
+        false
+    }
+
+    /// Can the window function be incrementally computed using
+    /// bounded memory?
+    ///
+    /// If this function returns true, [`Self::create_evaluator`] must
+    /// implement [`PartitionEvaluator::evaluate`]
+    fn supports_bounded_execution(&self) -> bool {
+        false
     }
```

Comment:
When `include_rank` flag is `true`,  `evaluate_with_rank_all` will be called. This method is basically same with `evaluate_all` in the spirit. (It takes all the data and produces all the output in single pass). However, since `evaluate_with_rank_all` requires additional arguments (such as rank boundaries). We do not unify their API, to not recalculate rank boundaries each time (even if we do not use them). 
Certainly, we can move this trait to `PartitionEvaluator` also. However, I thought this would be confusing. Hence didn't move it. (I will think about how to combine `evaluate_with_rank_all` and `evaluate_all` without calculating rank boundaries unnecessarily).

Maybe we can present to the user just a subset of the `PartitionEvaluator` methods. They wouldn't see `evaluate_with_rank_all` either (Just like your suggestion in [option 2](https://github.com/apache/arrow-datafusion/issues/5781#issuecomment-1583105449)).

Reply:
For anyone following along, I think we went with the `evaluate_with_rank_all` approach: https://github.com/apache/arrow-datafusion/blob/c3d5d77e447e51c2cca814a67706e5ab3e050ced/datafusion/physical-expr/src/window/partition_evaluator.rs#L209-L217